strawberryfields/tdm/tdmprogram.py

```diff
@@ -268,20 +387,139 @@ def __init__(self, N, name=None):
+                                "Program cannot be used with the device '{}' "
+                                "due to incompatible parameter.".format(device.target)
+                            )
+                # Now we will check explicitly if the parameters in the program match
+                num_symbolic_param = 0  # counts the number of symbolic variables, which are labeled consecutively by the context method
+
+                for k, param_name in enumerate(param_names):
+                    # Obtain the value of the corresponding parameter in the program
+                    program_param = self.rolled_circuit[i].op.p[k]
+
+                    # make sure that hardcoded parameters in the device layout are correct
+                    if not isinstance(param_name, str):
+                        if not program_param == param_name:
+                            raise CircuitError(
+                                "Program cannot be used with the device '{}' "
+                                "due to incompatible parameter. Parameter has value '{}'"
+                                "while its valid value is '{}'".format(
+                                    device.target, program_param, param_name
+                                )
+                            )
+                        continue
+
+                    # Obtain the relevant parameter range from the device
+                    param_range = device.gate_parameters[param_name]
+                    if sf.parameters.par_is_symbolic(program_param):
+                        # If it is a symbolic value go and lookup its corresponding list in self.tdm_params
+                        local_p_vals = self.tdm_params[num_symbolic_param]
+
+                        for x in local_p_vals:
+                            if not x in param_range:
+                                raise CircuitError(
+                                    "Program cannot be used with the device '{}' "
+                                    "due to incompatible parameter. Parameter has value '{}'"
+                                    "while its valid range is '{}'".format(
+                                        device.target, x, param_range
+                                    )
+                                )
+                        num_symbolic_param += 1
+
+                    else:
+                        # If it is a numerical value check directly
+                        if not program_param in param_range:
+                            raise CircuitError(
+                                "Program cannot be used with the device '{}' "
+                                "due to incompatible parameter. Parameter has value '{}'"
+                                "while its valid range is '{}'".format(
+                                    device.target, program_param, param_range
+                                )
+                            )
+        return self
```

Comment:
Should the logic be like the following?

```python
if compiler == "gaussian":
    return ...

if compiler == "TD2" and device is not None:
    ...
    return self

raise CircuitError("Unknown compiler etc.")
```

This avoids the case where `prog.compile()` or `prog.compile(compiler="TD3")` simply returns the existing program, which might not be intuitive.

Reply:
I like this! It makes sense. Although, if we decide to change `if compiler == "TD2" and device is not None:` to `if device is not None:`, there would need to be another error message.